lucene/misc/src/java/org/apache/lucene/misc/index/IndexRearranger.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.misc.index;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import org.apache.lucene.index.CodecReader;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.FilterCodecReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.IndexWriterConfig;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.NoMergePolicy;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.NamedThreadFactory;
+
+/**
+ * Copy and rearrange index according to document selectors, from input dir to output dir Length of
+ * documentSelectors determines how many segments there will be
+ */
+public class IndexRearranger {
+  protected final Directory input, output;
+  protected final IndexWriterConfig config;
+  protected final List<DocumentSelector> documentSelectors;
+
+  public IndexRearranger(
+      Directory input,
+      Directory output,
+      IndexWriterConfig config,
+      List<DocumentSelector> documentSelectors) {
```

Comment:
Maybe we could (followon PR maybe) have some sort of utility class implementing `DocumentSelector` by opening an existing index, crawling its docids to segments based on some specified primary key field or so?

Then this command-line tool could take a pointer to source index (to extract `DocumentSelector` per segment), name of primary key field, a pointer to input index to rewrite, and a pointer to filesystem where the rearranged index should be written?

Reply:
Good idea. I didn't include that part for keeping generality, but I think it is also worth implementing an example `DocumentSelector` as you suggested, I'll add one in this PR.

Reply:
Awesome, thanks for adding this!